src/Nethermind/Nethermind.Evm.Precompiles/Bls/BlsExtensions.cs

```diff
@@ -2,88 +2,83 @@
 
-        if (!ValidRawFp(raw[..BlsConst.LenFp]) || !ValidRawFp(raw[BlsConst.LenFp..]))
+        Result result;
+        if ((result = ValidRawFp(raw[..BlsConst.LenFp])) &&
+            (result = ValidRawFp(raw[BlsConst.LenFp..])))
         {
-            return false;
-        }
+            // set to infinity point by default
+            p.Zero();
 
-        // set to infinity point by default
-        p.Zero();
+            ReadOnlySpan<byte> fp0 = raw[BlsConst.LenFpPad..BlsConst.LenFp];
+            ReadOnlySpan<byte> fp1 = raw[(BlsConst.LenFp + BlsConst.LenFpPad)..];
 
-        ReadOnlySpan<byte> fp0 = raw[BlsConst.LenFpPad..BlsConst.LenFp];
-        ReadOnlySpan<byte> fp1 = raw[(BlsConst.LenFp + BlsConst.LenFpPad)..];
+            bool isInfinity = !fp0.ContainsAnyExcept((byte)0) && !fp1.ContainsAnyExcept((byte)0);
+            if (isInfinity)
+            {
+                return Result.Success;
+            }
 
-        bool isInfinity = !fp0.ContainsAnyExcept((byte)0) && !fp1.ContainsAnyExcept((byte)0);
-        if (isInfinity)
-        {
-            return true;
-        }
-
-        p.Decode(fp0, fp1);
-        if (!p.OnCurve())
-        {
-            return false;
+            p.Decode(fp0, fp1);
+            return p.OnCurve() ? Result.Success : Errors.G1PointSubgroup;
         }
 
-        return true;
+        return result;
     }
 
+
     // decodes and checks point is on curve
-    public static bool TryDecodeRaw(this G2 p, ReadOnlySpan<byte> raw)
+    public static Result TryDecodeRaw(this G2 p, ReadOnlySpan<byte> raw)
     {
         if (raw.Length != BlsConst.LenG2)
         {
-            return false;
-        }
-
-        if (!ValidRawFp(raw[..BlsConst.LenFp]) ||
-            !ValidRawFp(raw[BlsConst.LenFp..(2 * BlsConst.LenFp)]) ||
-            !ValidRawFp(raw[(2 * BlsConst.LenFp)..(3 * BlsConst.LenFp)]) ||
-            !ValidRawFp(raw[(3 * BlsConst.LenFp)..]))
-        {
-            return false;
+            return Errors.InvalidFieldLength;
         }
 
-        // set to infinity point by default
-        p.Zero();
-
-        ReadOnlySpan<byte> fp0 = raw[BlsConst.LenFpPad..BlsConst.LenFp];
-        ReadOnlySpan<byte> fp1 = raw[(BlsConst.LenFp + BlsConst.LenFpPad)..(2 * BlsConst.LenFp)];
-        ReadOnlySpan<byte> fp2 = raw[(2 * BlsConst.LenFp + BlsConst.LenFpPad)..(3 * BlsConst.LenFp)];
-        ReadOnlySpan<byte> fp3 = raw[(3 * BlsConst.LenFp + BlsConst.LenFpPad)..];
-
-        bool isInfinity = !fp0.ContainsAnyExcept((byte)0)
-            && !fp1.ContainsAnyExcept((byte)0)
-            && !fp2.ContainsAnyExcept((byte)0)
-            && !fp3.ContainsAnyExcept((byte)0);
-
-        if (isInfinity)
+        Result result;
+        if ((result = ValidRawFp(raw[..BlsConst.LenFp])) &&
+            (result = ValidRawFp(raw[BlsConst.LenFp..(2 * BlsConst.LenFp)])) &&
+            (result = ValidRawFp(raw[(2 * BlsConst.LenFp)..(3 * BlsConst.LenFp)])) &&
+            (result = ValidRawFp(raw[(3 * BlsConst.LenFp)..])))
```

Comment:
Reads pretty weird with multiple assignments in loop; why not assign outside and then just test `result`?

Is also more chars as 5 assignments vs 1

```suggestion
        Result result = ValidRawFp(raw[..BlsConst.LenFp]) &&
            ValidRawFp(raw[BlsConst.LenFp..(2 * BlsConst.LenFp)]) &&
            ValidRawFp(raw[(2 * BlsConst.LenFp)..(3 * BlsConst.LenFp)]) &&
            ValidRawFp(raw[(3 * BlsConst.LenFp)..]))
        if (result)
```

Reply:
This allows for lazy evaluation, while result is not `bool`. We also use result later to return the error message.
Maybe with custom `&&` operator on `Result`?

Reply:
Ok I came out with something